### DISCUSSION TITLE
(Java Agent) beforeMethod is triggered twice for a single request?

### DISCUSSION THREAD

**Discussion by biuld** [apache/skywalking#11771] [Q&A]
Here are my plugin definitions.
```java
public class SpringWebMVCInterceptorInstrumentation extends ClassInstanceMethodsEnhancePluginDefineV2 {

    @Override
    protected ClassMatch enhanceClass() {
        return NameMatch.byName("com.github.biuld.demo.service.impl.HelloServiceImpl");
    }

    @Override
    public ConstructorInterceptPoint[] getConstructorsInterceptPoints() {
        return null;
    }

    @Override
    public InstanceMethodsInterceptV2Point[] getInstanceMethodsInterceptV2Points() {
        return new InstanceMethodsInterceptV2Point[] {
                new DeclaredInstanceMethodsInterceptV2Point()   {
                    @Override
                    public ElementMatcher<MethodDescription> getMethodsMatcher() {
                        return ElementMatchers.named("sayHello");
                    }

                    @Override
                    public String getMethodsInterceptorV2() {
                        return "com.github.biuld.skywalking.apm.plugin.intercept.SpringWebMVCInterceptor";
                    }

                    @Override
                    public boolean isOverrideArgs() {
                        return false;
                    }
                }
        };
    }
}
```
```java
public class SpringWebMVCInterceptor implements InstanceMethodsAroundInterceptorV2 {


    int count = 0;

    @Override
    public void beforeMethod(EnhancedInstance objInst, Method method, Object[] allArguments, Class<?>[] argumentsTypes, MethodInvocationContext context) throws Throwable {
        count++;
        System.out.println("count: " + count);
        System.out.println(Arrays.toString(Thread.currentThread().getStackTrace()));
    }

    @Override
    public Object afterMethod(EnhancedInstance objInst, Method method, Object[] allArguments, Class<?>[] argumentsTypes, Object ret, MethodInvocationContext context) throws Throwable {
        return ret;
    }

    @Override
    public void handleMethodException(EnhancedInstance objInst, Method method, Object[] allArguments, Class<?>[] argumentsTypes, Throwable t, MethodInvocationContext context) {
        if (ContextManager.isActive()) {
            ContextManager.activeSpan().log(t);
        }
    }
}
```
```java
@Service
public class HelloServiceImpl implements HelloService {

    @Autowired
    private WorldService worldService;

    @Override
    public String sayHello() {
        return "hello, " + worldService.sayWorld();
    }
}
```

And here is the [console output](https://www.diffchecker.com/LYK2RW2a/), on the left is the first call and on the right is the second:
![image](https://github.com/apache/skywalking/assets/36144654/882f8a78-0c29-4564-a641-727a480b55b7)

Then I checked the lines showed by the diff:
![image](https://github.com/apache/skywalking/assets/36144654/8ad09c9d-e3d7-4f0d-88e1-f5ddb9960b4d)

It seems to me that zuper is not the original method rather than the enhanced one. I don't know if my code is wrong or bytebuddy is not correctly passing the `@SuperCall ` parameter (I guess it's because mutable reference passing around and got accidentally changed). Can anyone help?

**Comment by wu-sheng** [comment]
I don't follow your point. The auxiliary method is created. I can't see what you mean by your stacks.

**Reply by biuld** [reply]
My intercept point is org.springframework.web.servlet.DispatcherServlet#doDispatch which should be called once per request. And in beforeMethod, I count the times being called and prints the stacktrace. Found that beforeMethod was invoked twice.

**Reply by biuld** [reply]
The two printed stacktrace show little difference which tells me that zuper.call() is the second invocation of beforeMethod

**Reply by wu-sheng** [reply]
You count the invocation doesn't prove the issue. Because as a dispatcher, it could be called more than once during the call chain of HTTP handling. That is a common case.

If you want to confirm the mutiple invoke issue, use the simple method, rather than a framework.

**Reply by biuld** [reply]
But my endpoint is very simple that only returns "hello, world!"

**Reply by wu-sheng** [reply]
I mean, your stack doesn't show duplicated intercepting.

**Reply by biuld** [reply]
I'm not quite follow your word "duplicated intercepting". My question is that `zuper` seems not to be the original method but the enhanced one. And that could possibly explain why my console shows 2 numbers and two stack trace.

**Reply by wu-sheng** [reply]
How to generate method by doing intercepting, it is Bytebuddy project scope. This is not determined by us.

Interceptor is only being called once, that is for sure.

**Reply by biuld** [reply]
> Because as a dispatcher, it could be called more than once during the call chain of HTTP handling. That is a common case.

I would try a more simple method when I have time. But in my "hello,world" endpoint case, `doDispatch` should only be invoked once is pretty straightforward to me.

**Reply by biuld** [reply]
> Interceptor is only being called once

But if my guess is correct, the bytebuddy bug breaks this rule.

**Reply by wu-sheng** [reply]
I am not sure how you make this, but it can't be twice, because, if it is invoked twice, it will invoke endless in a dead loop. 
Could you follow the logic? So, your assumption is not correct.

**Reply by wu-sheng** [reply]
The only logical explanation is, the method is called in some conditions by the method itself with different conditions, and then exit.

**Reply by biuld** [reply]
> it will invoke endless in a dead loop.

Not exactly! Looking at the stacktrace I posted, the `intercept` method replace the original method. So the first invocation happens at line 57. Then it continues to line 67 which calls the enhanced instance , i.e. the `intercept` method again! It's like the original method is wrapped twice.

**Reply by biuld** [reply]
@wu-sheng I've updated my code above. Also, I can provide another relevant example. I've used the `@Trace` annotation, at my `sayHello` method and it creates two local spans. And I assume this would be unintended.
```java
@Service
public class HelloServiceImpl implements HelloService {

    @Autowired
    private WorldService worldService;

    @Override
    @Trace
    public String sayHello() {
        return "hello, " + worldService.sayWorld();
    }
}
```
![image](https://github.com/apache/skywalking/assets/36144654/96e3d4e6-7176-4f57-b29c-9e16b9e40d6d)

**Reply by wu-sheng** [reply]
Your case doesn't happen in our test scenario. 

This is how we use `@Trace` in the demo codes, https://github.com/apache/skywalking-java/blob/d5b99f9434530403569cba3a1d0fbd74e07448c9/test/plugin/scenarios/apm-toolkit-trace-scenario/src/main/java/test/apache/skywalking/apm/testcase/toolkit/controller/TestService.java#L61

And in the expectation file, we verified, this doesn't have nested span, https://github.com/apache/skywalking-java/blob/d5b99f9434530403569cba3a1d0fbd74e07448c9/test/plugin/scenarios/apm-toolkit-trace-scenario/config/expectedData.yaml#L33.

So generally, this is tested in our codebase, it only invoked once. 

We can't tell what happens in your private codebase, but the above test scenario is verifying the codes commit by commit. Please recheck what is different and unique in your test.

**Reply by wu-sheng** [reply]
Please note, I don't mean your description is wrong. My point is that it seems different from our test and understanding.

**Reply by biuld** [reply]
@wu-sheng Could you try my code above? My situation is not that hard to reproduce. I am using 8.9.0 java agent and 8.9.1 server and 2.7.18 spring boot. Just write a simple `helloworld` controller and test it.

**Reply by wu-sheng** [reply]
I am not using Java agent nowadays for my work. I may need hours to set up env back to work, but I don't have that time. 
But the auto-testing framework has been up and running for years, so it should be trustable. And 8.9 is a very old Java agent version.

**Reply by biuld** [reply]
Let alone the `@Trace` example. I've also change the intercept point from `org.springframework.web.servlet.DispatcherServlet#doDispatch` to `com.github.biuld.demo.service.impl.HelloServiceImpl#sayHello` and problem remains

**Reply by biuld** [reply]
If you don't have the luxury of reproducing it, does any other collaborator can help?

**Comment by biuld** [comment]
@wu-sheng Should open an issue for this problem? I've encountered the same problem with a more recent version of java agent and skywalking server.

**Reply by wu-sheng** [reply]
As I replied, we wouldn't take responsibility for your dev env.
So, this isn't a bug unless official plugins have similar issues to create duplicated spans.
So far, nothing liked this.

**Reply by wu-sheng** [reply]
Many traces and spans are collected on demo which is usinglatest version, so do by many OSS users and vendors.
No one reported duplicated span. 

If this exists at our end, I believe more people should show up with official use cases.
